### PR TITLE
Extension Detection

### PR DESCRIPTION
**Pull request by Gusted**
Pretty sure before we `add` any elements to the DOM just like scripts, we already inject the firstElementChild 

Resolves: #2227
Resolves: #2730

Not Tested

### CODE REVIEW

File: src/inject/dynamic-theme/index.ts [hunk 1]
```diff
@@ -329,12 +331,30 @@ function stopWatchingForUpdates() {
     stopWatchingForInlineStyles();
     removeDOMReadyListener(onDOMReady);
 }
+function checkMeta() {
```
Comment (alexanderby) [inline src/inject/dynamic-theme/index.ts:334]: Every function should have a single purpose. So I would split it into 2 functions: `isAnotherDarkReaderInstanceActive` and `createDarkReaderInstanceMarker` (maybe there's a better word than `Marker`). And also it should remove the marker, when `removeDynamicTheme` is called.

File: src/inject/dynamic-theme/index.ts [hunk 2]
```diff
@@ -329,12 +331,30 @@ function stopWatchingForUpdates() {
...
+    const meta: HTMLMetaElement = document.querySelector('meta[name="darkreader"]');
+    if (meta) {
+        if (meta.content !== UUID) {
+            return true;
+        }
+        return false;
+    } else {
+        const metaElement: HTMLMetaElement = document.createElement('meta');
+        metaElement.name = "darkreader";
```
Comment (alexanderby) [inline src/inject/dynamic-theme/index.ts:343]: code-style

File: src/inject/dynamic-theme/uuid.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// `hex` variable contain all available hex values into array.
+function hexify(number: Number) {
+    return number < 16 ? "0" + number.toString(16) : number.toString(16);
+}
+// Secure & Fast UUID generator implementation of https://tools.ietf.org/html/rfc4122 version 4.
+export function generateUUID() {
```
Comment (alexanderby) [inline src/inject/dynamic-theme/uuid.ts:6]: I've found out that this is invalid UUID. You can simply copy the last snippet from here https://stackoverflow.com/a/2117523/4137472

And another radical solution: we don't really need UUID, you can rename it to UID, simply generate 4 32bit numbers, convert them `toString(16)` and compose. That will have event less chance of collision than UUID and will not have weird dashes.

Also you can remove comments.

File: src/inject/dynamic-theme/index.ts [hunk 4]
```diff
@@ -14,9 +14,11 @@ import {getCSSFilterValue} from '../../generators/css-filter';
 import {modifyColor} from '../../generators/modify-colors';
 import {createTextStyle} from '../../generators/text-style';
 import {FilterConfig, DynamicThemeFix} from '../../definitions';
+import {generateUID} from './uuid';
 import {createAdoptedStyleSheetOverride, AdoptedStyleSheetManager} from './adopted-style-manger';
 
 const variables = new Map<string, string>();
+const UUID = generateUID();
```
Comment (alexanderby) [inline src/inject/dynamic-theme/index.ts:21]: I think it's worth renaming it to `INSTANCE_ID`.

File: src/inject/dynamic-theme/uuid.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+
+export function generateUID() {
+    const s = window.crypto.getRandomValues(new Uint8Array(16));
+    return s[0].toString(16) + s[1].toString(16) + s[2].toString(16) + s[3].toString(16) +
```
Comment (alexanderby) [inline src/inject/dynamic-theme/uuid.ts:4]: Everything was OK with your `hexify` function, except that you could make it shorter like `(x < 16 ? '0' : '') + x.toString(16)`. Now you can have a situation when `f + ff == ff + f`.

If you decide to make it not compatible with the spec, you can simply use `Array.from(...).map(hexify).join('') to not repeat the same expression 16 times.

File: src/inject/dynamic-theme/uuid.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+
+export function generateUID() {
+    const s = window.crypto.getRandomValues(new Uint8Array(16));
```
Comment (alexanderby) [inline src/inject/dynamic-theme/uuid.ts:3]: `window.` is unnecessary, `crypto` is a global object.

**Comment by Gusted** [comment]
BTW it's 80/20 for who is running first the check 80% it's the API so maybe their could be some improvements to let the Extension still handle as some have custom settings. Or we could have somehow getting the settings into the API from the extension so it will be actually efficienter.

**Comment by swrobel** [comment]
I'd like to propose an alternative solution (a bit late, I know, as this is already implemented), but hear me out.

There are a few issues with this solution:
1. It uses a custom meta tag that is specific to Dark Reader, rather than a standards compatible existing tag.
1. It uses a custom meta tag that is technically invalid, and thus won't render through some frameworks (like Ruby on Rails, as I've discovered). The tag includes an invalid `name` value, `darkreader`, which is not on the [official list](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta/name). This causes Rails to silently drop the tag when rendering the page, presumably since it deems it invalid.

Instead, I propose using `<meta name="color-scheme">` which you can read about by scrolling a short ways down [this page](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/meta/name). This is easy to check with javascript, ie: `document.querySelector('meta[name="color-scheme"]').content.includes("dark")`.

You can test this js on my site, https://metasurfforecast.com/ which is where I've been trying to use the existing solution to disable Dark Reader.

**Comment by Gusted** [comment]
> It uses a custom meta tag that is specific to Dark Reader, rather than a standards compatible existing tag.

This is, to check if any instance of Dark Reader is active on that web pages to not get conflicts with the Dark Reader API and let site disable Dark Reader on their site.

> It uses a custom meta tag that is technically invalid, and thus won't render through some frameworks (like Ruby on Rails, as I've discovered). The tag includes an invalid name value, darkreader, which is not on the official list. This causes Rails to silently drop the tag when rendering the page, presumably since it deems it invalid.

Oh, yeah that's quite unfortunate.

> Instead, I propose using <meta name="color-scheme"> which you can read about by scrolling a short ways down this page. This is easy to check with javascript, ie: document.querySelector('meta[name="color-scheme"]').content.includes("dark").

I think that could work. Just a question so I don't misinterpret the documentation. You can specify `dark light` inside the content of the meta. Whereby the documentation says about the following:
```
Indicating multiple color schemes indicates that the first scheme is preferred by the document, but that the second specified scheme is acceptable if the user prefers it.
```
What for me indicates the first value in content is being used as `default` value for a fresh new user.
As your site has `<meta content="light dark" name="color-scheme">` Whereby for me the default was actually a dark theme *not sure if it comes to the '(prefers-color-scheme: dark)' media Query*. 

I think a more globally solution which is reliable is actionable, thanks for the proposal!

**Comment by swrobel** [comment]
@Gusted thanks for the thorough response! You are correct that my site does render by default in light theme for users that don't have `prefers-color-scheme` support, or when their system color scheme is light. The dark theme will render when `prefers-color-scheme` is `dark` as you noticed (and as specified by the spec).

**Comment by Gusted** [comment]
@swrobel #3902

**Comment by alexanderby** [comment]
@swrobel What are you trying to achieve? Are you trying to disable Dark Reader for your website?

**Comment by swrobel** [comment]
> @swrobel What are you trying to achieve? Are you trying to disable Dark Reader for your website?

Yes, see https://github.com/darkreader/darkreader/pull/2855#issuecomment-713040345

**Comment by alexanderby** [comment]
OK, I've left some comments there.

**Comment by atwright147** [comment]
How about `<meta http-equiv="x-is-dark" content="true">`?

Prefixing with an `x-` is valid (see `<meta http-equiv="X-UA-Compatible" content="IE=9; IE=8; IE=7; IE=EDGE" />` as an example. This could also be set at a HTTP header level to work around the issue that Dark Reader might start before the `<head>` has started rendering.

**Comment by hadialqattan** [comment]
A workaround to disable Darkreader extension: https://github.com/hadialqattan/no-darkreader

**Comment by PeterHindes** [comment]
Hey btw found a bug. If you disable and then reenable dark reader, the meta tag is ignored.

**Pull request merged**